### ISSUE TITLE
IRkernel - cannot install repr

### ISSUE THREAD

**Issue by KatrinPechinger** [IRkernel/IRkernel#303]
I tried installing the R kernel with the code provided, but I get an error message saying " installation of package ‘repr’ had non-zero exit status". 
I've tried removing the IRkernel and IRdisplay packages and re- installing the packages in the correct order, as someone previously suggested, but this failed as well, as a timeout was reached.

Here's my R console output: 

> install.packages(c('repr', 'IRkernel', 'IRdisplay'),
+                  repos = c('http://irkernel.github.io/', getOption('repos')))

  There is a binary version available but the source version is later:
     binary source needs_compilation
repr    0.3    0.4             FALSE

trying URL 'http://irkernel.github.io/bin/windows/contrib/3.2/IRkernel_0.5.zip'
Content type 'application/zip' length 70035 bytes (68 KB)
downloaded 68 KB

trying URL 'http://irkernel.github.io/bin/windows/contrib/3.2/IRdisplay_0.3.zip'
Content type 'application/zip' length 23946 bytes (23 KB)
downloaded 23 KB

package ‘IRkernel’ successfully unpacked and MD5 sums checked
package ‘IRdisplay’ successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\hrakzp\AppData\Local\Temp\RtmpERZS3L\downloaded_packages
installing the source package ‘repr’

trying URL 'http://irkernel.github.io/src/contrib/repr_0.4.tar.gz'
Content type 'application/octet-stream' length 11355 bytes (11 KB)
downloaded 11 KB

'\\AKLDFS12\HOME$\hrakzp\My Documents\RStudio'
CMD.EXE was started with the above path as the current directory.
UNC paths are not supported.  Defaulting to Windows directory.
'\\AKLDFS12\HOME$\hrakzp\My' is not recognized as an internal or external command,
operable program or batch file.
Warning in install.packages :
  running command '"//AKLDFS12/HOME$/hrakzp/My Documents/R/R-3.2.3/bin/x64/R" CMD INSTALL -l "\\AKLDFS12\HOME$\HRAKZP\My Documents\R\R-3.2.3\library" C:\Users\hrakzp\AppData\Local\Temp\RtmpERZS3L/downloaded_packages/repr_0.4.tar.gz' had status 1
Warning in install.packages :
  installation of package ‘repr’ had non-zero exit status

The downloaded source packages are in
	‘C:\Users\hrakzp\AppData\Local\Temp\RtmpERZS3L\downloaded_packages’

> IRkernel::installspec()
Error in loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]]) : 
  there is no package called ‘repr’

remove.packages("IRkernel")
remove.packages("IRdisplay")

> devtools::install_github("IRkernel/repr")
Downloading GitHub repo IRkernel/repr@master
from URL https://api.github.com/repos/IRkernel/repr/zipball/master
Error in curl::curl_fetch_memory(url, handle = handle) : 
  Timeout was reached

----------------------------------------------------------------------------------------------------------------------------

Any help would be greatly appreciated.

System specs: Windwos 7 Professional
R version 3.2.3

**Comment by janschulz** [comment]
> Defaulting to Windows directory.
'\AKLDFS12\HOME$\hrakzp\My' is not recognized as an internal or external command,
operable program or batch file.

This looks like R can't handle your setup with the sapace in one of the paths your R wants to use?

> Error in curl::curl_fetch_memory(url, handle = handle) : 
Timeout was reached

Ok, I've no clue what is happening here. This looks like github is not reachable from your place? Or can't download zip files? Can you access https://api.github.com/repos/IRkernel/repr/zipball/master in your browser?

**Comment by KatrinPechinger** [comment]
Yes, I can access the zip-file ( https://api.github.com/repos/IRkernel/repr/zipball/master ) in my browser. Not sure why R can't pull it.

**Comment by gacek91** [comment]
Same thing here guys, I believe the problem is on your side - I can install all the packages but IRKernel.

**Comment by janschulz** [comment]
@gacek91 What is your output?

@KatrinPechinger I still think the problem in your case is your setup: your default lib path points to an UNC path and it seems R CMD build/install can't handle that (or better it tried to put cmd there and that can't). Try setting R_USER to something on your local computer (eg this in a commandline to change it permanently to `C:\portabel\R` `setx R_USER C:\portabel\R` -> New packages will be installed in that location in version specific subdirs... You have to move all your currently installed packages there or reinstall).

**Comment by gacek91** [comment]
I used the command below:
```
install.packages(c('rzmq','repr','IRkernel','IRdisplay'),
                 repos = c('http://irkernel.github.io/', getOption('repos')))
```
That's the output:
```
* installing *source* package ‘rzmq’ ...
** libs
clang++ -std=c++11 -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG -I../inst/cppzmq -I/usr/local/include -I/usr/local/include/freetype2 -I/opt/X11/include    -fPIC  -Wall -mtune=core2 -g -O2 -c interface.cpp -o interface.o
interface.cpp:22:10: fatal error: 'zmq.h' file not found
#include <zmq.h>
         ^
1 error generated.
make: *** [interface.o] Error 1
ERROR: compilation failed for package ‘rzmq’
* removing ‘/Library/Frameworks/R.framework/Versions/3.3/Resources/library/rzmq’
Warning in install.packages :
  installation of package ‘rzmq’ had non-zero exit status
* installing *source* package ‘IRdisplay’ ...
** R
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded
* DONE (IRdisplay)
ERROR: dependency ‘rzmq’ is not available for package ‘IRkernel’
* removing ‘/Library/Frameworks/R.framework/Versions/3.3/Resources/library/IRkernel’
Warning in install.packages :
  installation of package ‘IRkernel’ had non-zero exit status
```

So instead I tried:
```
library("devtools")
install_github("IRkernel/repr")
install_github("IRkernel/IRdisplay")
install_github("IRkernel/IRkernel") # should install a zmq lib from cran
```

And here's the output:

```
> library("devtools")
> install_github("IRkernel/repr")
Downloading GitHub repo IRkernel/repr@master
from URL https://api.github.com/repos/IRkernel/repr/zipball/master
Installing repr
'/Library/Frameworks/R.framework/Resources/bin/R' --no-site-file --no-environ --no-save --no-restore --quiet CMD  \
  INSTALL  \
  '/private/var/folders/kg/pydc_kpx0kd0j009w7c35s2c0000gn/T/RtmpLDlIq7/devtools26414c4fad72/IRkernel-repr-bd5e9b6'  \
  --library='/Library/Frameworks/R.framework/Versions/3.3/Resources/library' --install-tests 

* installing *source* package ‘repr’ ...
** R
** tests
** preparing package for lazy loading
Failed with error:  ‘there is no package called ‘Cairo’’
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded
* DONE (repr)
> install_github("IRkernel/IRdisplay")
Downloading GitHub repo IRkernel/IRdisplay@master
from URL https://api.github.com/repos/IRkernel/IRdisplay/zipball/master
Installing IRdisplay
Skipping 1 package ahead of CRAN: repr
'/Library/Frameworks/R.framework/Resources/bin/R' --no-site-file --no-environ --no-save --no-restore --quiet CMD  \
  INSTALL  \
  '/private/var/folders/kg/pydc_kpx0kd0j009w7c35s2c0000gn/T/RtmpLDlIq7/devtools26412e39e0ae/IRkernel-IRdisplay-69fd5d8'  \
  --library='/Library/Frameworks/R.framework/Versions/3.3/Resources/library' --install-tests 

* installing *source* package ‘IRdisplay’ ...
** R
** tests
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded
* DONE (IRdisplay)
> install_github("IRkernel/IRkernel") # should install a zmq lib from cran
Downloading GitHub repo IRkernel/IRkernel@master
from URL https://api.github.com/repos/IRkernel/IRkernel/zipball/master
Installing IRkernel
Skipping 1 unavailable package: IRdisplay
Skipping 1 package ahead of CRAN: repr
'/Library/Frameworks/R.framework/Resources/bin/R' --no-site-file --no-environ --no-save --no-restore --quiet CMD  \
  INSTALL  \
  '/private/var/folders/kg/pydc_kpx0kd0j009w7c35s2c0000gn/T/RtmpLDlIq7/devtools26411154192c/IRkernel-IRkernel-a939fe9'  \
  --library='/Library/Frameworks/R.framework/Versions/3.3/Resources/library' --install-tests 

* installing *source* package ‘IRkernel’ ...
** R
** inst
** preparing package for lazy loading
Error : .onLoad failed in loadNamespace() for 'pbdZMQ', details:
  call: dyn.load(file, DLLpath = DLLpath, ...)
  error: nie można załadować współdzielonego obiektu  '/Library/Frameworks/R.framework/Versions/3.3/Resources/library/pbdZMQ/libs/pbdZMQ.so':
  dlopen(/Library/Frameworks/R.framework/Versions/3.3/Resources/library/pbdZMQ/libs/pbdZMQ.so, 6): Library not loaded: /Builds/CRAN-QA-Simon/packages/mavericks-x86_64/Rlib/3.3/pbdZMQ/libs/libzmq.4.dylib
  Referenced from: /Library/Frameworks/R.framework/Versions/3.3/Resources/library/pbdZMQ/libs/pbdZMQ.so
  Reason: image not found
ERROR: lazy loading failed for package ‘IRkernel’
* removing ‘/Library/Frameworks/R.framework/Versions/3.3/Resources/library/IRkernel’
BŁĄD: Command failed (1)
```

**Comment by flying-sheep** [comment]
we have new (temporary) instructions coming in: IRkernel/irkernel.github.io#29

check out the already updated [README](https://github.com/IRkernel/IRkernel#installation).

CC @janschulz

**Comment by gacek91** [comment]
Still can't get through with IRkernel
```
* installing *source* package ‘IRkernel’ ...
** R
** inst
** preparing package for lazy loading
Error : .onLoad failed in loadNamespace() for 'pbdZMQ', details:
  call: dyn.load(file, DLLpath = DLLpath, ...)
  error: nie można załadować współdzielonego obiektu  #####says it can't can't load shared object
'/Library/Frameworks/R.framework/Versions/3.3/Resources/library/pbdZMQ/libs/pbdZMQ.so':
  dlopen(/Library/Frameworks/R.framework/Versions/3.3/Resources/library/pbdZMQ/libs/pbdZMQ.so, 6): Library not loaded: /Builds/CRAN-QA-Simon/packages/mavericks-x86_64/Rlib/3.3/pbdZMQ/libs/libzmq.4.dylib
  Referenced from: /Library/Frameworks/R.framework/Versions/3.3/Resources/library/pbdZMQ/libs/pbdZMQ.so
  Reason: image not found
ERROR: lazy loading failed for package ‘IRkernel’
* removing ‘/Library/Frameworks/R.framework/Versions/3.3/Resources/library/IRkernel’
BŁĄD: Command failed (1)
```

**Comment by flying-sheep** [comment]
can you reinstall pbdZMQ?

**Comment by gacek91** [comment]
Did twice:

```
Error : .onLoad nie powiodło się w funkcji 'loadNamespace()' dla pakietu 'pbdZMQ', szczegóły:
  wywołanie: dyn.load(file, DLLpath = DLLpath, ...)
  błąd:  nie można załadować współdzielonego obiektu  ##again can't load '/Library/Frameworks/R.framework/Versions/3.3/Resources/library/pbdZMQ/libs/pbdZMQ.so':
  dlopen(/Library/Frameworks/R.framework/Versions/3.3/Resources/library/pbdZMQ/libs/pbdZMQ.so, 6): Library not loaded: /Builds/CRAN-QA-Simon/packages/mavericks-x86_64/Rlib/3.3/pbdZMQ/libs/libzmq.4.dylib
  Referenced from: /Library/Frameworks/R.framework/Versions/3.3/Resources/library/pbdZMQ/libs/pbdZMQ.so
  Reason: image not found
BŁĄD: ładowanie pakietu lub przestrzeni nazw nie powiodło się dla pakietu ‘pbdZMQ’ #BŁĄD obviously means ERROR.

> 
```

**Comment by takluyver** [comment]
pbdZMQ may not make those install issues go away as neatly as we had hoped :-(

**Comment by flying-sheep** [comment]
i thought it ships a bundled version of it that gets used :disappointed:

it really brings out the linux snobbery in me if windows and OSX *constantly* have issues

**Comment by takluyver** [comment]
Did we figure out what the issue is with pbdZMQ and/or what can be done?

**Comment by janschulz** [comment]
On windows, it works: it includes a copy of `libzmq.dll`. Also just had a look into https://cran.r-project.org/bin/macosx/contrib/r-release/pbdZMQ_0.1-1.tgz and it also inlcudes a `libzmq.dylib`, so I think that works too... 

@gacek91 can you do a `remove.packages("pbdZMQ")` followed by a `install.packages("pbdZMQ")` and then `devtools::install_github("irkernel/IRkernel")` and post the complete output if that doesn't work?

**Comment by janschulz** [comment]
CC: @snoweye can you tell us if the R3.3 Mac OS X package on CRAN is ok or not? The above 

> Library not loaded: /Builds/CRAN-QA-Simon/packages/mavericks-x86_64/Rlib/3.3/pbdZMQ/libs/libzmq.4.dylib

suggests that something is wrong here? 

Unfortunately, I don't have a OSX system to check if the lib has unresolved symbols :-(

**Comment by janschulz** [comment]
BTW: we have the same problem on Travis with Mac OS X: https://github.com/IRkernel/IRkernel/pull/330

**Comment by gacek91** [comment]
```
> devtools::install_github("irkernel/IRkernel")
Downloading GitHub repo irkernel/IRkernel@master
from URL https://api.github.com/repos/irkernel/IRkernel/zipball/master
Installing IRkernel
Skipping 1 unavailable package: IRdisplay
Installing 1 package: repr
próbowanie adresu URL 'https://cran.rstudio.com/bin/macosx/mavericks/contrib/3.3/repr_0.7.tgz'
Content type 'application/x-gzip' length 54819 bytes (53 KB)
==================================================
downloaded 53 KB


The downloaded binary packages are in
	/var/folders/kg/pydc_kpx0kd0j009w7c35s2c0000gn/T//RtmpwgyEs5/downloaded_packages
'/Library/Frameworks/R.framework/Resources/bin/R' --no-site-file --no-environ --no-save  \
  --no-restore --quiet CMD INSTALL  \
  '/private/var/folders/kg/pydc_kpx0kd0j009w7c35s2c0000gn/T/RtmpwgyEs5/devtools139375fb213/IRkernel-IRkernel-de32d2a'  \
  --library='/Library/Frameworks/R.framework/Versions/3.3/Resources/library'  \
  --install-tests 

* installing *source* package ‘IRkernel’ ...
** R
** inst
** preparing package for lazy loading
Error : .onLoad failed in loadNamespace() for 'pbdZMQ', details:
  call: dyn.load(file, DLLpath = DLLpath, ...)
  error: nie można załadować współdzielonego obiektu  '/Library/Frameworks/R.framework/Versions/3.3/Resources/library/pbdZMQ/libs/pbdZMQ.so':
  dlopen(/Library/Frameworks/R.framework/Versions/3.3/Resources/library/pbdZMQ/libs/pbdZMQ.so, 6): Library not loaded: /Builds/CRAN-QA-Simon/packages/mavericks-x86_64/Rlib/3.3/pbdZMQ/libs/libzmq.4.dylib
  Referenced from: /Library/Frameworks/R.framework/Versions/3.3/Resources/library/pbdZMQ/libs/pbdZMQ.so
  Reason: image not found
ERROR: lazy loading failed for package ‘IRkernel’
* removing ‘/Library/Frameworks/R.framework/Versions/3.3/Resources/library/IRkernel’
BŁĄD: Command failed (1)
```

**Comment by janschulz** [comment]
I've opened an issue at https://github.com/snoweye/pbdZMQ/issues/28

**Comment by flying-sheep** [comment]
```
> devtools::install_github("irkernel/IRkernel")
Downloading GitHub repo irkernel/IRkernel@master
from URL https://api.github.com/repos/irkernel/IRkernel/zipball/master
Installing IRkernel
Skipping 1 unavailable package: IRdisplay
```

you need to install IRdisplay first anyway. and for somebody `devtools::install_github("irkernel/IRkernel")` tried to install the wrong package, so please @janschulz start to recommend using the correct capitalization: `"IRkernel/IRkernel"`

**Comment by janschulz** [comment]
On my system, this is not true:

```
> devtools::install_github("irkernel/irkernel")
Downloading GitHub repo irkernel/irkernel@master
from URL https://api.github.com/repos/irkernel/irkernel/zipball/master
Installing IRkernel
Skipping 1 unavailable package: IRdisplay
"C:/PROGRA~1/R/R-3.3/bin/x64/R" --no-site-file --no-environ --no-save --no-restore --quiet CMD  \
  INSTALL  \
  "C:/Users/jschulz/AppData/Local/Temp/Rtmp2Jfwyf/devtools3cf476135914/IRkernel-IRkernel-cb7fd61"  \
  --library="C:/portabel/R/R/win-library/3.3" --install-tests 

* installing *source* package 'IRkernel' ...
** R
** inst
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded
*** arch - i386
*** arch - x64
* DONE (IRkernel)
> 
```

**Comment by flying-sheep** [comment]
should be system independent. and it was true for one person over at IRdisplay. unless you can show me where the GitHub API says its URLs are case insensitive, we better take no risks.

**Comment by janschulz** [comment]
The help page seems to suggest that they are case sensitive: https://help.github.com/articles/error-repository-not-found/#check-your-spelling Not sure why it never mattered on my end...

**Comment by takluyver** [comment]
I've just tried cloning with both SSH and HTTPS, and it doesn't seem to be case sensitive. This is on Linux. Perhaps the help page is out of date...

**Comment by flying-sheep** [comment]
well, fact is that using the correct case will always work, so let’s just give that to people.

**Comment by flying-sheep** [comment]
the temporary install instructions on https://irkernel.github.io should work for our three packages.

pbdZMQ not working has to be resolved at snoweye/pbdZMQ#28

**Comment by janschulz** [comment]
According to snoweye/pbdZMQ#28 this is now fixed on devel and release, so if you are sill on oldrelease (3.2), I'm not sure if it will be fixed?
